File: mpp.network/src/main/java/server/ServerConnectionManager.java
package server;

import connection.RequestListener;
import connection.RequestListenerProtocol;
import observer.ObserverServerProtocol;
import response.NotificationType;
import response.ResponseNotification;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;

/**
 * Name:        {ClassName}
 * Effect:      {ClassEffect}
 * Date:        02/04/2017
 * Tested:      False
 *
 * @author Alexandru Stoica
 * @version 1.0
 */

public class ServerConnectionManager implements ConcurrentServerProtocol {

    private ServerSocket serverSocket;
    private Integer port;
    private ObserverServerProtocol observer;
    private ArrayList<RequestListenerProtocol> clients;

    public ServerConnectionManager(Integer port) {
        this.port = port;
        this.clients = new ArrayList<>();
    }

    public void start() {
        try {
            serverSocket = new ServerSocket(port);
            while (true) {
                Socket socket = serverSocket.accept();
                processClient(socket);
            }
        } catch (IOException error) {
            System.out.println(error.getMessage());
        }
    }

    public void setObserver(ObserverServerProtocol observer) {
        this.observer = observer;
    }

    public void processClient(Socket socket) {
        Thread thread = createThread(socket);
        thread.start();
    }

    public void notifyClients(NotificationType notification) {
        clients.forEach(client ->
                client.sendNotification(new ResponseNotification(notification)));
    }

    public void stop() {
        try {
            serverSocket.close();
        } catch (IOException error) {
            System.out.print(error.getMessage());
        }
    }

    public Thread createThread(Socket socket) {
        RequestListenerProtocol listener = new RequestListener(socket);
        listener.setObserver(observer);
        clients.add(listener);
        return new Thread(listener);
    }

}

File: Database Management Systems/Event Database Manager/NetworkComponent/Connection/ServerConnectionManager.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using NetworkComponent.Listener;
using NetworkComponent.Response;
using NetworkComponent.Transferable;

namespace NetworkComponent.Connection
{
    public class ServerConnectionManager: IServerConnection
    {
        /// <summary>
        /// The port used by the server's system.
        /// </summary>
        private readonly int _port;

        /// <summary>
        /// The host (localhost) used by the server's system.
        /// </summary>
        private readonly string _host;

        /// <summary>
        /// Checks if the server is running.
        /// </summary>
        private bool _isRunning;

        /// <summary>
        /// The list of clients what are currently connected to the server.
        /// </summary>
        public List<IRequestListener> Clients;
        
        /// <summary>
        /// Manages the server's connection to it's clients.
        /// </summary>
        /// <param name="host"> The server's port </param>
        /// <param name="port"> The server's host (localhost) </param>
        public ServerConnectionManager(string host, int port)
        {
            _host = host;
            _port = port;
            Clients = new List<IRequestListener>();
        }

        /// <summary>
        /// Starts the server's activity by waiting & connecting to it's clients. 
        /// </summary>
        public void Start()
        {
            _isRunning = true;
            Console.WriteLine("Connection Starts with port: " + _port);
            var address = IPAddress.Parse(_host);
            var endPoint = new IPEndPoint(address, _port);
            var server = new TcpListener(endPoint);
            server.Start();
            while (_isRunning)
            {
                Console.WriteLine("Waiting For Clients ...");
                var tcpClient = server.AcceptTcpClient();
                Console.WriteLine("Client Connected ...");
                ProcessClient(tcpClient.GetStream());
            }
        }
        /// <summary>
        /// Runs the client's task in a new thread. 
        /// (listens for client's requests)
        /// </summary>
        /// <param name="stream"> The system's socket </param>
        private void ProcessClient(NetworkStream stream)
        {
            var client = CreateThread(stream);
            client.Start();
        }

        private Thread CreateThread(NetworkStream stream)
        {
            IRequestListener listener = new RequestListener(stream);
            listener.Subscribe(this);
            Clients.Add(listener);
            return new Thread(listener.Start);
        }

        public void Stop()
        {
            Console.Write("Connection Ends with port: " + _port);
            _isRunning = false;
        }

        public void Update(ITransferable notification)
        {
            foreach (var client in Clients) 
            {
                client.Send((IResponse)notification);
            }
        }
    }
}
